src/python/zquantum/core/hamiltonian.py

```diff
@@ -69,3 +71,124 @@ def group_comeasureable_terms_greedy(
+        expecval: An ExpectationValues object containing the expectation
+                  values of the operators and their squares. Optionally, contains
+                  values of operator products to compute covariances.
+    Returns:
+        frame_variances: A Numpy array of the computed variances for each frame
+    """
+
+    if expecval is None:
+        frame_variances = [ np.sum(np.array(list(x.terms.values()))**2) for x in groups ] # Covariances are ignored; Variances are set to 1
+    else:
+        group_sizes = np.array([ len(g.terms.keys()) for g in groups ])
+        assert np.sum(group_sizes) == len(expecval.values)
+        real_expecval = expectation_values_to_real(expecval)
+        pauli_variances = 1. - real_expecval.values**2
+        frame_variances = []
+        for i, group  in enumerate(groups):
+            coeffs = np.array(list(group.terms.values()))
+            offset = 0 if i == 0 else np.sum(group_sizes[:i])
+            pauli_variances_for_group = pauli_variances[offset : offset + group_sizes[i]]
+            frame_variances.append(np.sum(coeffs**2 * pauli_variances_for_group))
+
+    return np.array(frame_variances)
+
+def get_expectation_values_from_rdms(interactionrdm: InteractionRDM, 
+        qubitoperator: QubitOperator, 
+        sort_terms: bool = False) -> ExpectationValues:
+    """ Computes expectation values of a qubitOperator
+        given a fermionic interactionRDM operator from
+        OpenFermion.
+
+        Args:
+            interactionrdm: interaction RDM to use for the expectation values
+                computation, as an OF InteractionRDM object
+            qubitoperator: qubit operator to compute the expectation values for
+                in the form of an OF QubittOperator object
+            sort_terms: whether or not the input qubit operator needs to be sorted before calculating expectations
+        Returns:
+            expectations: expectation values of Pauli strings in the qubit operator
+    """
+    if sort_terms:
+        terms_iterator = sorted(
+            qubitoperator.terms.items(), key=lambda x: abs(x[1]), reverse=True
+        )
+    else:
+        terms_iterator = qubitoperator.terms.items()
+    reordered_qubitoperator = QubitOperator()
+    for term, coefficient in terms_iterator:
+        reordered_qubitoperator += QubitOperator(term, coefficient)
+
+    expectations_packed = interactionrdm.get_qubit_expectations(reordered_qubitoperator)
+
+    if () in expectations_packed.terms:
+        del expectations_packed.terms[()] # Expectation of the constant term is excluded from expectation values
+
+    expectations = np.real(np.array(list(expectations_packed.terms.values()))) # should we add an assert to catch large Im parts
+    # Clip expectations if they fell out of [-1 , 1] due to numerical errors
+    expectations[expectations < -1.0] = -1.0
+    expectations[expectations >  1.0] =  1.0
```

Comment:
I don't know if you are familiar with it, but there's `np.clip` function that does precisely that but can handle both edges of the interval at the same time. I think the usage here would look like this: `np.clip(expectations, -1, 1, out=expectations)`. Using `out` keyword will perform the operation in place, which would be consistent with the current code.

Reply:
I replaced my code with the `np.clip` call (I never used it before, it's handy! Thank you!)